metadata-ingestion/src/datahub/utilities/file_backed_collections.py

```diff
@@ -255,16 +257,44 @@ def __delitem__(self, key: str) -> None:
             raise KeyError(key)
 
     def __iter__(self) -> Iterator[str]:
+        # Cache should be small, so safe list cast to avoid mutation during iteration
+        cache_keys = list(self._active_object_cache.keys())
+        yield from cache_keys
+
         cursor = self._conn.execute(f"SELECT key FROM {self.tablename}")
         for row in cursor:
-            if row[0] in self._active_object_cache:
-                # If the key is in the active object cache, then SQL isn't the source of truth.
-                continue
+            if row[0] not in cache_keys:
+                yield row[0]
+
+    def items(self) -> Iterator[Tuple[str, _VT]]:
+        # Implement manually to avoid __getitem__ calls
+        cache_keys = list(self._active_object_cache.keys())
+        yield from self._active_object_cache.items()
+
+        cursor = self._conn.execute(f"SELECT key, value FROM {self.tablename}")
+        for row in cursor:
+            if row[0] not in cache_keys:
+                yield row[0], self.deserializer(row[1])
 
-            yield row[0]
+    def filtered_items(self, cond_sql: str) -> Iterator[Tuple[str, _VT]]:
+        """
+        Flush the cache and iterate through a filtered list of the dictionary's items.
 
-        for key in self._active_object_cache:
-            yield key
+        Args:
+            cond_sql: Conditional expression for WHERE statement, e.g. `x = 0 AND y = "value"`
+
+        Returns:
+            Iterator of filtered (key, value) pairs.
+        """
+        self.flush()
+        for row in self._conn.execute(
+            f"SELECT key, value FROM {self.tablename} WHERE {cond_sql}"
+        ):
+            yield row[0], self.deserializer(row[1])
```

Comment:
same as above - are we requiring no mutations to values during iteration?

Reply:
Hmm I thought we were requiring no mutations at all. Mutations during iteration is cool but it feels dangerous to me -- it works for most cases, but it seems too easy to do something out of the ordinary and get burned, e.g.:

```
@dataclass
class MyObject:
    name: string
    value: int
    parents: List[string]

my_dict = FileBackedDict[MyObject]()
...
important_entry = MyObject["main"]
for k, obj in my_dict:
    if obj.name == important_entry.name:
        important_entry.parents.append(k)
```
Here, we'll catch some but not all of the `important_entry` mutations, once `important_entry` is cache busted

Or even worse...
```    
for k, obj in my_dict.items():
    for parent in obj.parents:
        my_dict[parent].value += obj.value
    obj.parents = []
```
If `len(obj.parents) > max_cache_size` we won't catch the mutation in the last line

Reply:
in that case, we should rename to `items_no_mutation` or something and then the constraint will be more clear in the calling code

I'm just trying to prevent incorrect usage of the interface because it'll become super confusing if bugs happen

Reply:
fwiw I'm ok with requiring no mutations during iteration

Reply:
I'd prefer not to do the renames -- I understand wanting to prevent incorrect usage, but I think we should hopefully be able to say a blanket "you cannot store objects in a FileBackedDict and mutate them" and not require different names. Otherwise, I don't think renaming to `items_no_mutation` is sufficient -- it's not just when iterating through `items` that you can't mutate. You can't mutate during any iteration, so you shouldn't be able to do `for key in my_file_backed_dict`, nor can you mutate during regular key access, e.g. `if key in my_file_backed_dict: ...`. I don't think it's practical to rename all of these usages

Reply:
Ah, are you saying we won't catch changes to keys during iteration? e.g.

```
d = FileBackedDict()
d['a'] = 'a'
d['b'] = 'b'
for k,v in d.items():
    d['b'] = 'd'
    print(k,v)
```
And then if we iterate 'a' first, we should get the update in the next loop?

I do think the `items()` spec is supposed to pick up these changes, which we won't right now